engine/src/main/java/org/terasology/engine/world/chunks/pipeline/ChunkProcessingPipeline.java

```diff
@@ -44,120 +32,162 @@
  */
 public class ChunkProcessingPipeline {
 
-    private static final int NUM_TASK_THREADS = 4;
+    private static final int NUM_TASK_THREADS = 2;
+    private static final int CHUNKS_AT_ONCE = 8;
     private static final Logger logger = LoggerFactory.getLogger(ChunkProcessingPipeline.class);
 
     private final List<ChunkTaskProvider> stages = Lists.newArrayList();
-    private final Thread reactor;
-    private final CompletionService<Chunk> chunkProcessor;
-    private final ThreadPoolExecutor executor;
+
     private final Function<Vector3ic, Chunk> chunkProvider;
     private final Map<Vector3ic, ChunkProcessingInfo> chunkProcessingInfoMap = Maps.newConcurrentMap();
-    private int threadIndex;
+    private final List<Subscription> subs = new ArrayList<>();
+    private final Scheduler scheduler;
+
+    private final Object completeSignal = new Object();
 
     /**
      * Create ChunkProcessingPipeline.
      */
-    public ChunkProcessingPipeline(Function<Vector3ic, Chunk> chunkProvider, Comparator<Future<Chunk>> comparable) {
+    public ChunkProcessingPipeline(Function<Vector3ic, Chunk> chunkProvider, Flux<Chunk> chunkStream) {
         this.chunkProvider = chunkProvider;
 
-        executor = new ThreadPoolExecutor(
-                NUM_TASK_THREADS,
-                NUM_TASK_THREADS, 0L,
-                TimeUnit.MILLISECONDS,
-                new PriorityBlockingQueue(800, unwrappingComporator(comparable)),
-                this::threadFactory,
-                this::rejectQueueHandler) {
-            @Override
-            protected <T> RunnableFuture<T> newTaskFor(Callable<T> callable) {
-                RunnableFuture<T> newTaskFor = super.newTaskFor(callable);
-                return new PositionFuture<>(newTaskFor, ((PositionalCallable) callable).getPosition());
-            }
-        };
-        chunkProcessor = new ExecutorCompletionService<>(executor,
-                new PriorityBlockingQueue<>(800, comparable));
-        reactor = new Thread(this::chunkTaskHandler);
-        reactor.setDaemon(true);
-        reactor.setName("Chunk-Processing-Reactor");
-        reactor.start();
+        scheduler = Schedulers.newParallel("chunk processing", NUM_TASK_THREADS);
+        Flux<Chunk> stream = chunkStream.subscribeOn(scheduler);
+        for (int i = 0; i < NUM_TASK_THREADS; i++) {
```

Comment:
use the parallel scheduler this is cpu bounded. we probably also want to limit the number of subscribers to the number of cores on the machine.
```suggestion
        Flux<Chunk> stream = chunkStream.subscribeOn(Schedulers.parallel());
        for (int i = 0; i < NUM_TASK_THREADS; i++) {
```

Reply:
I am using the parallel scheduler, but with a name and limiting it to `NUM_TASK_THREADS` threads. From the docs, it sounds like that's the only difference between `newParallel` and `parallel`. More than `NUM_TASK_THREADS` workers wouldn't be helpful as there are only that many subscriptions, so are you just talking about the case where the host CPU has less than `NUM_TASK_THREADS` cores?

Reply:
just realized this is a consuming operation so maybe push the operation to boundedElastic(). should re-claim the thread when we're finished.

Reply:
Do you mean when the player is standing still and chunk generation has finished, or after a world is closed? In the former case, I don't think that usually happens for very long, so I would guess that reclaiming the workers would have more overhead; in the latter case, the whole scheduler is already disposed after the world is closed.